conveyor/adapter/example.go

```diff
@@ -0,0 +1,290 @@
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+package adapter
+
+import (
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/insolar/insolar/conveyor/queue"
+	"github.com/insolar/insolar/log"
+	"github.com/pkg/errors"
+)
+
+type processElement struct {
+	elementID   idType
+	handlerID   idType
+	taskPayload interface{}
+	respSink    PulseConveyorSlotResponseSink
+	cancelInfo  *cancelInfoT
+}
+
+type cancelInfoT struct {
+	id     uint64
+	cancel chan bool
+	flush  chan bool
+}
+
+func newCancelInfo(id uint64) *cancelInfoT {
+	return &cancelInfoT{
+		id:     id,
+		cancel: make(chan bool, 1),
+		flush:  make(chan bool, 1),
+	}
+}
+
+type taskHolderT struct {
+	taskHolderLock sync.Mutex
+	tasks          map[uint32][]*cancelInfoT
+}
+
+func newTaskHolder() taskHolderT {
+	return taskHolderT{
+		tasks: make(map[uint32][]*cancelInfoT),
+	}
+}
+
+func (th *taskHolderT) add(info *cancelInfoT, pulseNumber uint32) {
+	log.Infof("[ taskHolderT.add ] Adding pulseNumber: %d. Id: %d", pulseNumber, info.id)
+	th.taskHolderLock.Lock()
+	defer th.taskHolderLock.Unlock()
+
+	el, ok := th.tasks[pulseNumber]
+	if !ok {
+		th.tasks[pulseNumber] = []*cancelInfoT{info}
+	} else {
+		th.tasks[pulseNumber] = append(el, info)
+	}
+}
+
+func processStop(cancelList []*cancelInfoT, flush bool) {
+	for _, el := range cancelList {
+		if flush {
+			log.Info("[ processStop ] flush: ", el.id)
+			el.flush <- true
+		} else {
+			log.Info("[ processStop ] cancel: ", el.id)
+			el.cancel <- true
+		}
+	}
+}
+
+func (th *taskHolderT) stop(pulseNumber uint32, flush bool) {
+	log.Infof("[ taskHolderT.stop ] Stopping pulseNumber: %d, flush: %d", pulseNumber, flush)
+	th.taskHolderLock.Lock()
+	defer th.taskHolderLock.Unlock()
+
+	cancelList, ok := th.tasks[pulseNumber]
+	if !ok {
+		log.Info("[ taskHolderT.stop ] No such pulseNumber: ", pulseNumber)
+		return
+	}
+
+	processStop(cancelList, flush)
+
+	delete(th.tasks, pulseNumber)
+}
+
+func (th *taskHolderT) stopAll(flush bool) {
+	th.taskHolderLock.Lock()
+	defer th.taskHolderLock.Unlock()
+
+	log.Infof("[ taskHolderT.stopAll ] flush: ", flush)
+
+	for _, cancelList := range th.tasks {
+		processStop(cancelList, flush)
+	}
+
+	th.tasks = make(map[uint32][]*cancelInfoT)
+
+}
+
+// SimpleWaitAdapter holds all adapter logic
+type SimpleWaitAdapter struct {
+	queue             queue.IQueue
+	processingStarted uint32
+	stopProcessing    uint32
+	processingStopped chan bool
+	// adapterID comes from configuration
+	adapterID uint32
+
+	taskHolder taskHolderT
+}
+
+type simpleWaitAdapterInputData struct {
+	waitPeriodMilliseconds int
+}
+
+// NewSimpleWaitAdapter creates new instance of SimpleWaitAdapter
+func NewSimpleWaitAdapter() PulseConveyorAdapterTaskSink {
```

Comment:
Adapter must implement `PulseConveyorSlotResponseSink` interface - adapter pushes response/nested event to slot response queue, and Slot must implement `PulseConveyorAdapterTaskSink` interface - push task to adapter queue.